Enable Bolt 3.0 and 4.0

majensen opened the pull request:
This PR enables a large portion of Bolt protocol versions 3.0 and 4.0. Additional functionality is also added to the shell client to take advantage of explicit transactions (3.0+ `:begin` `:commit` and `:rollback`) and of having multiple databases (4.0+ `:dbname`).

majensen committed:
Update README.md

hey this is a fork

majensen committed:
bolt 3.0 - add messages and update negotiation

majensen committed:
bolt-3.0 - building out transaction machinery WIP

transaction.h
transaction.c
define tx object that is analogous to result_stream object

majensen committed:
feat(bolt 3.0) - begin, run, rollback, commit

* tx object filled out
* api-level functions:
  * neo4j_begin_tx(connection,timeout,mode)
  * neo4j_run_in_tx(tx, statement, params) - easy, just use result_stream.c functions
  * neo4j_commit_tx(tx)
  * neo4j_rollback_tx(tx)

* receive callbacks for each of begin, commit, rollback
* new NEO4J_TRANSACTION_FAILED errno

Some working decisions made on how to store server argvs in tx
structure. Leaving as neo4j_* types; getters could convert to strings

* Need to destroy tx, presume that should happen under the hood in commit and
rollback. If tx is gone, tho, need to communicate final status back to
user some other way.

majensen committed:
feat(bolt 3.0) - compiles

New messages (HELLO, GOODBYE, BEGIN, COMMIT, ROLLBACK) all included at
this point.

To do: deal with RESET substituting for ACK_FAILURE in 3.0

majensen committed:
feat(bolt 3.0) - old checks pass

in ./tests

majensen committed:
feat(bolt 3.0) - txn code added to build; compiles

but it's WIP

adhering better to @cleishm's object patterns

majensen committed:
feat(bolt-3.0) : added check_transaction

to tests; passes.

majensen committed:
feat(bolt-3.0) rm cruft

majensen committed:
feat(bolt-3.0): test run stmt in tx

majensen committed:
feat(bolt-3.0): modify INIT argv for bolt 3

(if bolt 2, the old arguments are preserved)
This gives a successful connection to neo4j:3.5

majensen committed:
feat(bolt-3.0): doc for txn api functions

majensen committed:
feat(bolt-3.0): extra dict required in RUN argv

even if it's empty, for bolt 3+

majensen committed:
Update README.md

highlight this as a fork

majensen committed:
feat(bolt-3.0): RESET in place of ACK_FAILURE

when version >= 3

majensen committed:
feat(bolt-3.0): catch expired tx in run

catch expired transaction error in run results
and update the transaction.

majensen committed:
Merge branch 'feat-bolt-3.0' of https://github.com/majensen/libneo4j-client into feat-bolt-3.0

majensen committed:
feat(bolt-3.0) - update dir struct to refect master

majensen committed:
Merge branch 'feat-bolt-3.0'

majensen committed:
feat(bolt-3.0) bolt 3.0 into master

make check passes

majensen committed:
feat(shell 3.0) - a start on adding tx cmds

WIP - not working yet

majensen committed:
feat(bolt 3.0): Add begin/rollback/commit to shell

Add functioning :begin, :commit, :rollback commands

majensen committed:
feat(bolt-4.0) handle maj/min versions, dbname

Manage protocol major and minor versions in protocol negotiation.
Attach database name to config object - preserves the neo4j_run API

majensen committed:
feat(bolt 4.0): enable 4.0 database selection

Most of this has to do with enabling the RUN message to accept 3
arguments (statement, parameters, and the "extra" arg) rather than
2. The extra arg is a neo4j_map that can transmit the desired database
name in the "db" key value.

A :dbname command is added to the shell.

Negotiation is updated to accept v4.

bolt v4 is not fully implemented yet.

majensen committed:
feat(bolt 4.0) - this should work, but doesn't

Several messages (RUN, PULL, DISCARD) have an additional or
"extra" map argument in 4.0. This should not be a problem for
the machinery of neo4j-client. It is a matter of
* creating a map with the required infomation
* setting the next available slot in the request _argv[] array to
  this map
* setting the argument count req->argc to the new size of _argv =
  argv.

Such a message queues ok; logging at TRACE level indicates that the
arguments are intact. However, when the request comes up for sending
(via connection.c:send_requests()) -- that is, when it is pulled off
the request queue in the connection object -- the new map is trashed.
This is evident via an error from the serializer (called in
neo4j_connection_send), that indicates that a map key is not a
neo4j_string -- even though that map was successfully printed by the
logger in neo4j_session_run(), when the request was created/enqueued.

Sample log output against a real running 4.0 server:
``
TRACE [connection]: enqu BEGIN (0x7fdcc3e002a0) in 0x7fdcc3d200b0
DEBUG [connection]: sent BEGIN (0x7fdcc3e002a0) in 0x7fdcc3d200b0
DEBUG [connection]: rcvd SUCCESS in response to BEGIN (0x7fdcc3e002a0)
TRACE [connection]: enqu RUN{"CREATE (n:booga {thing:1})", {}, {db:"neo4j"}} (0x7fdcc3e00380) in 0x7fdcc3d200b0
TRACE [connection]: enqu PULL {n:-1,qid:-1} (0x7fdcc3e00460) in 0x7fdcc3d200b0
DEBUG [connection]: sent RUN (0x7fdcc3e00380) in 0x7fdcc3d200b0
ERROR [connection]: Error sending message on 0x7fdcc3d200b0: Map contains key of non-String type
TRACE [connection]: draining RUN (0x7fdcc3e00380) from queue on 0x7fdcc3d200b0
TRACE [connection]: draining PULL_ALL (0x7fdcc3e00460) from queue on 0x7fdcc3d200b0
Segmentation fault: 11
``

majensen committed:
Fix #1 - remove default timeout

Now set transation timeout value to 0 to inhibit timeout, set to -1
to fallback to server-side config timeout set.

majensen committed:
Merge branch 'master' into feat-bolt-4.0

majensen committed:
feat(bolt 4.0): this works

also, checks work for all versions; leaving version set to 4.

majensen committed:
feat(bolt 4.0): manage dbname

Add slot in shell state object for dbname and manage it there (not
config)
Remove dbname from config
Make dbname a noop for a version 3 server (rather than barfing)

majensen committed:
rm cruft

majensen committed:
update README

majensen committed:
rm cruft

majensen committed:
rm cruft

majensen committed:
Fix #3 #4 #6

Set DEFAULT_TIMEOUT in shell code to -1.
Provide static mem to the dbname string in shell state struct.
Added :reconnect command for simple reconnection after server enters failure
Might also fix #5?

majensen committed:
Fix #8 - reset properly

This was a fairly deep logical bug issue the original code -- which
probably didn't matter before there were transactions.

The basic problem is that the reset method short circuits if the
connection is marked as failed -- but that is precisely when you want
to attempt a reset.

To be fair -- the ACK_FAILURE / RESET process of Bolt 1 might have
worked differently, so that the original logic made sense.

majensen committed:
add txn features and protect from roll/comm closed txns

majensen committed:
fix bug in transaction.c

majensen committed:
workaround unusual bug

Refering to messages.h/.c:
In libneo4j-client, message types (BEGIN, FAILURE, SUCCESS, etc.) are
represented by constant pointers to static structures of the form
{ .name, .struct_signature }
where the name is the Bolt all-caps name as a string constant, and
the signature is the 1-byte code that represents that msg in the
protocol.

Care is taken to ensure that these pointers do not move, so that a
type is tested as (e.g.)
  type == NEO4J_SUCCESS_MESSAGE
where the rhs is #define'd as the constant pointer value.

Strangely (to me), these pointers are not stable when using the
library in Perl XS functions (specifically, with the library in
Neo4j::Client and the XS in Neo4j::Bolt).

Nevertheless, the contents of the structures are ok, so that if you
check a type by looking at the name,
  strcmp(type->name, "SUCCESS")
you always get what you want (i.e., only the addresses seem to move
around, but the data is not stomped on).

To preserve the original code for standalone compilation, I've wrapped
each message type test in
doing the string check in each case if -DNEOCLIENT_BUILD.

majensen committed:
complete unusual bug fix

majensen committed:
Fix #9 - also make :reconnect actually work

majensen committed:
fix bug introduced by fixing bug

and so on ad inifinitum

majensen commented:
Looking forward to a review of this work- thanks!

majensen commented:
@cleishm thinking about having a look?

cleishm commented:
Definitely thinking about it! ;-) Sorry, been very busy with my day job lately. I'll try to peek at it this week.

Once we get this reviewed and merged, we should chat about opening this repo up to other maintainers as well - like yourself?

majensen commented:
@cleishm would be an honor! I would have to actually learn C then--

majensen committed:
feat(new types): values.c/values.h type machinery

add structure based type definitions, constructors,
and accessors

majensen committed:
feat(new types): values.c/values.h type machinery

add structure based type definitions, constructors,
and accessors

majensen committed:
feat(new types): print functions

majensen committed:
feat(new types): print functions

majensen committed:
fix wrong assumptions; add some tests (WIP)

majensen committed:
fix wrong assumptions; add some tests (WIP)

majensen committed:
localtime test

majensen committed:
localtime test

majensen committed:
localtime test works

majensen committed:
localtime test works

majensen committed:
localtime datetime checks

majensen committed:
localtime datetime checks

majensen committed:
localdatetime

majensen committed:
localdatetime

majensen committed:
point2d point3d

majensen committed:
point2d point3d

majensen committed:
feat(v5.0): update supported versions

update supported versions in negotiation
add NEO4J_ELEMENTID type throughout the type system
- an element id is treated as a neo4j_string under the hood

majensen committed:
feat(v5.0): update supported versions

update supported versions in negotiation
add NEO4J_ELEMENTID type throughout the type system
- an element id is treated as a neo4j_string under the hood

majensen committed:
fix print/values for new structures

majensen committed:
fix print/values for new structures

majensen committed:
update deserialization tests for v5

majensen committed:
update deserialization tests for v5

majensen committed:
more

majensen committed:
more

majensen committed:
catch structs with known signatures, create correct types

majensen committed:
render durations acc to ISO 8601

majensen committed:
fixing time renderings

to coincide with Neo4j browser's

majensen committed:
update version offer

in version negotiation - enable all minor versions for v4 and v5
by offering ranges per
https://neo4j.com/docs/bolt/current/bolt/handshake/#bolt-version43

majensen committed:
Merge branch 'cleishm:main' into master

majensen committed:
Merge branch 'master' into feat-neo4j-5.0

majensen committed:
resolve

majensen committed:
update connection test

with new supported version string

majensen committed:
Merge pull request #11 from majensen/feat-neo4j-5.0

Feat neo4j 5.0 - all tests pass

majensen committed:
fix(supported versions): handshake exclude v5.1

which causes an obscure error at the server

majensen committed:
struct datatypes fix: legacy datetime fix

majensen committed:
feat(supported versions): dynamic versions

remove hard-coded supported versions
version_spec structure to describe versions
config struct contains supported version list
protocol negotiation uses config version list

majensen committed:
feat(supported versions): put version spec in config

config structure holds version specification array
parse simple version string comma-separated list to set
add config accessor to set via version string

majensen committed:
feat(supported versions): get and set in config

added options to shell - versions, --support to take a version string
to customize client-server handshake

majensen committed:
Update README.md
